### PR TITLE
chore: migrate Dropdown to TS

### PR DESCRIPTION
**Pull request by Mila2999**
fixes https://github.com/mondaycom/monday-ui-react-core/issues/1622

#### Basic

- [ ] Used plop (`npm run plop`) to create a new component.
- [ ] PR has description.
- [ ] New component is functional and uses Hooks.
- [ ] Component is written in TypeScript.

#### Style

- [ ] Styles are added to `NewComponent.modules.scss` file inside the `NewComponent` folder.
- [ ] Component uses CSS Modules.
- [ ] Design is compatible with [Monday Design System](https://design.monday.com/).
- [ ] Styles are defined using [monday-ui-style](https://github.com/mondaycom/monday-ui-react-core/tree/monorepo-style/packages/monday-ui-style) tokens
- [ ] Displays correctly in all the themes

#### Storybook

- [ ] Stories were added to `/src/components/NewComponent/__stories__/NewComponent.mdx` file.
- [ ] Stories include all flows of using the component.
- [ ] Stories implemented using [vibe-storybook-components](https://github.com/mondaycom/vibe-storybook-components) components and tokens.

#### Tests

- [ ] Tests are compliant with [TESTING_README.md](TESTING_README.md) instructions.

#### Accessibility

- [ ] Component is accessible.
- [ ] Component is compatible with [Vibe Accessibility Guidelines](https://style.monday.com/?path=/docs/foundations-accessibility--docs)

### CODE REVIEW

**Review by talkor — COMMENTED** [review]
Thank you @Mila2999! That's really great. Let's just avoid using `any` and leverage TS and use the actual types

**Comment by Mila2999** [comment]
thanks @talkor, i will work on fixing the types

**Comment by Mila2999** [comment]
Hi @talkor , i removed all any references except `ref={ref as React.RefObject<any>}` . any idea what can i assign to ref?

File: packages/core/src/components/Dropdown/Dropdown.tsx [hunk 1]
```diff
@@ -18,16 +17,284 @@ import {
...
+import { IComboboxOption } from "../Combobox/components/ComboboxConstants";
+import { Option } from "react-select/src/filters";
+import { HtmlAttributes } from "csstype";
+import { RefObject } from "preact";
+
+interface CustomSingleValueProps extends SingleValueProps<DropdownDefaultValue> {
+  Renderer: React.ComponentType;
+  readOnly: boolean;
+  selectedOption: DropdownDefaultValue;
+  singleVueWrapperClassName?: string;
```
Comment (talkor) [inline packages/core/src/components/Dropdown/Dropdown.tsx:39]: What is this property?

File: packages/core/src/components/Dropdown/DropdownConstants.js [hunk 2]
```diff
@@ -1,22 +0,0 @@
...
-export const DROPDOWN_MENU_POSITION = {
-  ABSOLUTE: "absolute",
-  FIXED: "fixed"
-};
-
-export const DROPDOWN_MENU_PLACEMENT = {
-  TOP: "top",
-  BOTTOM: "bottom",
-  AUTO: "auto"
-};
```
Comment (talkor) [inline packages/core/src/components/Dropdown/DropdownConstants.js:22]: Since we are yet to support using types externally correctly we don't want to break existing consumers' usages when consuming these objects from the dist folder, so we'll have to keep them intact until next major version

File: packages/core/src/components/Dropdown/DropdownConstants.tsx [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+}
+
+export enum DROPDOWN_MENU_PLACEMENT {
+  TOP = "top",
+  BOTTOM = "bottom",
+  AUTO = "auto"
+}
+
+export type DropdownDefaultValue = {
+  label: string;
```
Comment (talkor) [inline packages/core/src/components/Dropdown/DropdownConstants.tsx:33]: I think a better naming would be Option or something similar, as `default` implies it's a value and not a type

File: packages/core/src/components/Dropdown/Dropdown.tsx [hunk 4]
```diff
@@ -18,16 +17,284 @@ import {
...
+}
+
+interface CustomMenuProps extends MenuProps<Option, boolean> {
+  Renderer: React.ComponentType;
```
Comment (talkor) [inline packages/core/src/components/Dropdown/Dropdown.tsx:43]: This is coming from MenuProps, no?

File: packages/core/src/components/Dropdown/Dropdown.tsx [hunk 5]
```diff
@@ -18,16 +17,284 @@ import {
...
+  dropdownMenuWrapperClassName: string;
+}
+
+interface CustomOptionProps extends OptionProps<Option, boolean> {
+  Renderer: React.ComponentType;
```
Comment (talkor) [inline packages/core/src/components/Dropdown/Dropdown.tsx:48]: Same

File: packages/core/src/components/Dropdown/Dropdown.tsx [hunk 6]
```diff
@@ -177,44 +450,57 @@ const Dropdown = forwardRef(
     }, [size, rtl, insideOverflowContainer, insideOverflowWithTransformContainer, extraStyles, multi, multiline]);
 
     const Menu = useCallback(
-      props => (
+      (props: CustomMenuProps) => (
```
Comment (talkor) [inline packages/core/src/components/Dropdown/Dropdown.tsx:453]: I wonder if we better use the MenuComponent's interface already, it's out of this scope so maybe leaving it as any for now would be better. Same for the rest sub components here

File: packages/core/src/components/Dropdown/Dropdown.tsx [hunk 7]
```diff
@@ -177,44 +450,57 @@ const Dropdown = forwardRef(
...
 
     const onOptionRemove = useMemo(() => {
-      return function (optionValue, e) {
+      return function (optionValue: number, e: React.MouseEvent | React.KeyboardEvent) {
         if (customOnOptionRemove) {
-          customOnOptionRemove(selectedOptionsMap[optionValue]);
+          customOnOptionRemove((selectedOptionsMap as unknown as DropdownDefaultValue[])[optionValue]);
         }
-        const newSelectedOptions = selectedOptions.filter(option => option.value !== optionValue);
+        const newSelectedOptions = Array.isArray(selectedOptions)
```
Comment (talkor) [inline packages/core/src/components/Dropdown/Dropdown.tsx:500]: Is this because of typing issues? Isn't it always an array?

File: packages/core/src/components/Dropdown/Dropdown.tsx [hunk 8]
```diff
@@ -18,16 +17,284 @@ import {
...
+   */
+  extraStyles?: (...args: unknown[]) => unknown;
+  /**
+   * Maximum height of the menu before scrolling
+   */
+  maxMenuHeight?: number;
+  /**
+   * Tab index for keyboard navigation purposes
+   */
+  tabIndex?: number;
```
Comment (talkor) [inline packages/core/src/components/Dropdown/Dropdown.tsx:209]: We also support strings so it will break existing usages

File: packages/core/src/components/Dropdown/Dropdown.tsx [hunk 9]
```diff
@@ -140,34 +410,37 @@ const Dropdown = forwardRef(
         insideOverflowWithTransformContainer
       });
 
+      type BaseStyles = typeof baseStyles;
```
Comment (talkor) [inline packages/core/src/components/Dropdown/Dropdown.tsx:413]: Why is it defined here?

File: packages/core/src/components/Dropdown/Dropdown.tsx [hunk 10]
```diff
@@ -18,16 +17,284 @@ import {
...
+   */
+  openMenuOnClick?: boolean;
+  /**
+   * If set to true, clear button will be added
+   */
+  clearable?: boolean;
+  /**
+   * custom option render function
+   */
+  optionRenderer?: (option: IComboboxOption) => JSX.Element;
```
Comment (talkor) [inline packages/core/src/components/Dropdown/Dropdown.tsx:140]: Why using the ComboboxOption?

File: packages/core/src/components/Dropdown/Dropdown.tsx [hunk 11]
```diff
@@ -18,16 +17,284 @@ import {
...
+  rtl?: boolean;
+  /**
+   * Set default selected value
+   */
+  defaultValue?: DropdownDefaultValue[];
+  /**
+   * The component's value.
+   * When passed, makes this a [controlled](https://reactjs.org/docs/forms.html#controlled-components) component.
+   */
+  value?: DropdownDefaultValue[];
```
Comment (talkor) [inline packages/core/src/components/Dropdown/Dropdown.tsx:173]: Should also support single value no?

File: packages/core/src/components/Dropdown/Dropdown.tsx [hunk 12]
```diff
@@ -18,16 +17,284 @@ import {
...
+  /**
+  Pass closeMenuOnSelect to close the multi choose any time an options is chosen.
+  */
+  closeMenuOnSelect?: boolean;
+  // Won't be needed once we upgrade to react-select ^5.5 https://github.com/JedWatson/react-select/issues/4088#issuecomment-1276835389
+  /**
+   * If menu should be closed on scroll - helpful for some tricky use cases
+   * @default false, but true when insideOverflowContainer or insideOverflowWithTransformContainer are true
+   */
+  closeMenuOnScroll?: ((event: React.FocusEvent) => boolean) | boolean;
```
Comment (talkor) [inline packages/core/src/components/Dropdown/Dropdown.tsx:237]: Isn't it a boolean?

File: packages/core/src/components/Dropdown/Dropdown.tsx [hunk 13]
```diff
@@ -18,16 +17,284 @@ import {
...
+  tabSelectsValue?: boolean;
+  /**
+   * Overrides the build-in search filter logic - https://react-select.com/advanced#custom-filter-logic
+   * createFilter function is available at Dropdown.createFilter
+   */
+  filterOption?: (option: unknown, inputValue: string) => boolean;
 
-const Dropdown = forwardRef(
+  withReadOnlyStyle?: boolean;
+  OptionRenderer?: (option: IComboboxOption) => JSX.Element;
```
Comment (talkor) [inline packages/core/src/components/Dropdown/Dropdown.tsx:289]: Why ComboboxOption

**Review by talkor — CHANGES_REQUESTED** [review]
Thank you for taking the time to work on this! And apologies for the long delay. I wrote some comments and some notes to make this migration without breaking existing usages
Thanks!

File: packages/core/src/components/Dropdown/Dropdown.tsx [hunk 14]
```diff
@@ -18,16 +17,284 @@ import {
   ADD_AUTO_HEIGHT_COMPONENTS,
   defaultCustomStyles,
   DROPDOWN_CHIP_COLORS,
-  DROPDOWN_ID,
+  DROPDOWN_MENU,
```
[the remark this answers is not in the document]
Comment (Mila2999) [reply packages/core/src/components/Dropdown/Dropdown.tsx:20]: moved DROPDOWN_ID to DROPDOWN_MENU enum to be used like DROPDOWN_MENU.ID

File: packages/core/package.json [hunk 15]
```diff
@@ -119,7 +119,7 @@
     "react-transition-group": "^4.4.5",
     "react-virtualized-auto-sizer": "^1.0.7",
     "react-window": "^1.8.7",
-    "react-windowed-select": "^2.0.4",
+    "react-windowed-select": "3.1.2",
```
[the remark this answers is not in the document]
Comment (Mila2999) [reply packages/core/package.json:122]: in 3.1.1 react windowed select added typescript types, that's why i migrated it
image

File: packages/core/src/components/Dropdown/Dropdown.tsx [hunk 16]
```diff
@@ -18,16 +17,284 @@ import {
...
 } from "./DropdownConstants";
 import generateBaseStyles, { customTheme } from "./Dropdown.styles";
 import Control from "./components/Control/Control";
 import menuStyles from "./components/menu/menu.module.scss";
 import styles from "./Dropdown.module.scss";
+import { VibeComponent, VibeComponentProps } from "src/types";
+import { IComboboxOption } from "../Combobox/components/ComboboxConstants";
+import { Option } from "react-select/src/filters";
+import { HtmlAttributes } from "csstype";
+import { RefObject } from "preact";
```
[the remark this answers is not in the document]
Comment (Mila2999) [reply packages/core/src/components/Dropdown/Dropdown.tsx:33]: removed

File: packages/core/src/components/Dropdown/Dropdown.tsx (hunk 9 above)
Comment (Mila2999) [reply packages/core/src/components/Dropdown/Dropdown.tsx:413]: there is a Dropdown.styles.js file.
as I understand this is not a part of the current migration but I still need to use types somehow
so I used typeof to get style types.
this should be removed after migrating that file.

image

File: packages/core/src/components/Dropdown/Dropdown.tsx (hunk 11 above)
Comment (Mila2999) [reply packages/core/src/components/Dropdown/Dropdown.tsx]: you right, changed

File: packages/core/src/components/Dropdown/Dropdown.tsx (hunk 13 above)
Comment (Mila2999) [reply packages/core/src/components/Dropdown/Dropdown.tsx:284]: I took it from Combobox.tsx,
changing to DropdownOption

File: packages/core/src/components/Dropdown/Dropdown.tsx [hunk 17]
```diff
@@ -18,16 +17,284 @@ import {
...
+  DROPDOWN_MENU_POSITION,
+  DropdownDefaultValue
 } from "./DropdownConstants";
 import generateBaseStyles, { customTheme } from "./Dropdown.styles";
 import Control from "./components/Control/Control";
 import menuStyles from "./components/menu/menu.module.scss";
 import styles from "./Dropdown.module.scss";
+import { VibeComponent, VibeComponentProps } from "src/types";
+import { IComboboxOption } from "../Combobox/components/ComboboxConstants";
+import { Option } from "react-select/src/filters";
```
[the remark this answers is not in the document]
Comment (Mila2999) [reply packages/core/src/components/Dropdown/Dropdown.tsx:31]: removed and used OptionTypeBase

File: packages/core/src/components/Dropdown/DropdownConstants.tsx [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+  value: string | number | Theme;
```
[the remark this answers is not in the document]
Comment (Mila2999) [reply packages/core/src/components/Dropdown/DropdownConstants.tsx]: removed

File: packages/core/src/components/Dropdown/Dropdown.tsx [hunk 19]
```diff
@@ -18,16 +17,284 @@ import {
...
+  clearable?: boolean;
+  /**
+   * custom option render function
+   */
+  optionRenderer?: (option: IComboboxOption) => JSX.Element;
+  /**
+   * custom value render function
+   */
+  valueRenderer?: (...args: unknown[]) => unknown;
+  ValueRenderer?: (...args: unknown[]) => unknown;
```
[the remark this answers is not in the document]
Comment (Mila2999) [reply packages/core/src/components/Dropdown/Dropdown.tsx:140]: It seems there are two values in the forwardRef, and I prefer not to alter the existing code

File: packages/core/src/components/Dropdown/Dropdown.tsx (hunk 1 above)
Comment (Mila2999) [reply packages/core/src/components/Dropdown/Dropdown.tsx]: removed, not needed

File: packages/core/src/components/Dropdown/DropdownConstants.js (hunk 2 above)
Comment (Mila2999) [reply packages/core/src/components/Dropdown/DropdownConstants.js:22]: i understand, i will revert this change

File: packages/core/src/components/Dropdown/DropdownConstants.tsx (hunk 3 above)
Comment (Mila2999) [reply packages/core/src/components/Dropdown/DropdownConstants.tsx:33]: renamed to DropdownOption

File: packages/core/src/components/Dropdown/Dropdown.tsx (hunk 4 above)
Comment (Mila2999) [reply packages/core/src/components/Dropdown/Dropdown.tsx]: you right, removed and it work great

File: packages/core/src/components/Dropdown/Dropdown.tsx (hunk 5 above)
Comment (Mila2999) [reply packages/core/src/components/Dropdown/Dropdown.tsx]: removed Renderer

File: packages/core/src/components/Dropdown/Dropdown.tsx (hunk 7 above)
Comment (Mila2999) [reply packages/core/src/components/Dropdown/Dropdown.tsx:495]: there is a default value that is not an array, so I added this to check.
and yes there was a typing issue.

File: packages/core/src/components/Dropdown/Dropdown.tsx (hunk 8 above)
Comment (Mila2999) [reply packages/core/src/components/Dropdown/Dropdown.tsx:204]: changed to number | string.

File: packages/core/src/components/Dropdown/Dropdown.tsx (hunk 10 above)
Comment (Mila2999) [reply packages/core/src/components/Dropdown/Dropdown.tsx:135]: copyed from combobox file, changed to DropdownOption, i assume it should be dropdown option

File: packages/core/src/components/Dropdown/Dropdown.tsx [hunk 20]
```diff
@@ -18,16 +17,284 @@ import {
...
+   */
+  disabled?: boolean;
+  /**
+   * If set to true, dropdown won't be editable
+   */
+  readOnly?: boolean;
+  /**
+   * Called when menu is opened
+   */
+  onMenuOpen?: () => void;
```
[the remark this answers is not in the document]
Comment (Mila2999) [reply packages/core/src/components/Dropdown/Dropdown.tsx:83]: i added event: React.MouseEvent as param

File: packages/core/src/components/Dropdown/Dropdown.tsx [hunk 21]
```diff
@@ -18,16 +17,284 @@ import {
...
+  /**
+   * Called when menu is closed
+   */
+  onMenuClose?: () => void;
```
[the remark this answers is not in the document]
Comment (Mila2999) [reply packages/core/src/components/Dropdown/Dropdown.tsx:87]: i added event: React.MouseEvent as param

**Comment by Mila2999** [comment]
Hi @talkor , thank you for your time in reviewing my PR. I have addressed your comments and made the necessary fixes to the code

File: packages/core/package.json (hunk 15 above)
[the remark this answers is not in the document]
Comment (talkor) [reply packages/core/package.json:122]: There's a major bump in react-select in this library between those two versions (see [this](https://github.com/jacobworrel/react-windowed-select/compare/2.0.3-beta.0...ef0ed2e43d989347a1ef7d9bb4219e7247517cd0#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R28)), this might lead to issues as it's a depenedency

File: packages/core/types/packages.d.ts
Comment (YossiSaadi) [inline packages/core/types/packages.d.ts:1]: is there a real need in this one?
image
seem like `react-windowed-select` does have TS support at this version

File: packages/core/src/components/Dropdown/Dropdown.tsx [hunk 22]
```diff
@@ -20,14 +27,278 @@ import {
...
+   */
+  asyncOptions?: (inputValue: string) => Promise<DropdownOption[]>;
+  /**
+   * If set to true, fetched async options will be cached
+   */
+  cacheOptions?: boolean;
+  /**
+   * If set, `asyncOptions` will be invoked with its value on mount and the resolved results will be loaded
+   */
+  defaultOptions?: boolean | Record<string, string>[];
```
Comment (YossiSaadi) [inline packages/core/src/components/Dropdown/Dropdown.tsx:193]: Sounds like should be something like:
```
defaultOptions?: GroupedOptionsType<OptionTypeBase> | OptionsType<OptionTypeBase> | boolean;
```

Comment (YossiSaadi) [reply packages/core/src/components/Dropdown/Dropdown.tsx:193]: Everywhere where we use `DropdownOption`, we should be caution with it
We'll might need to replace to `OptionTypeBase` as a first step, and in the future make the Dropdown component accept a generic for `OptionType` that extends `OptionTypeBase`.

For now, let's leave it as is

File: packages/core/src/components/Dropdown/Dropdown.tsx [hunk 23]
```diff
@@ -20,14 +27,278 @@ import {
...
+  defaultValue?: DropdownOption[];
+  /**
+   * The component's value.
+   * When passed, makes this a [controlled](https://reactjs.org/docs/forms.html#controlled-components) component.
+   */
+  value?: DropdownOption[] | DropdownOption;
+  /**
+   * Select menu size from `Dropdown.size` - Dropdown.sizes.LARGE | Dropdown.sizes.MEDIUM | Dropdown.sizes.SMALL
+   */
+  size?: BaseSizes;
```
Comment (YossiSaadi) [inline packages/core/src/components/Dropdown/Dropdown.tsx:181]: Should be `SIZES`

File: packages/core/src/components/Dropdown/Dropdown.tsx [hunk 24]
```diff
@@ -397,13 +687,13 @@ Dropdown.defaultProps = {
...
   searchable: true,
   options: [],
-  menuPlacement: Dropdown.menuPlacements.BOTTOM,
-  menuPosition: Dropdown.menuPositions.ABSOLUTE,
+  menuPlacement: DROPDOWN_MENU_PLACEMENT.BOTTOM,
+  menuPosition: DROPDOWN_MENU_POSITION.ABSOLUTE,
   noOptionsMessage: NOOP,
   clearable: true,
-  size: SIZES.MEDIUM,
+  size: BaseSizes.MEDIUM,
```
Comment (YossiSaadi) [inline packages/core/src/components/Dropdown/Dropdown.tsx:694]: Should be `SIZES.MEDIUM`

File: packages/core/src/components/Dropdown/Dropdown.tsx [hunk 25]
```diff
@@ -397,13 +687,13 @@ Dropdown.defaultProps = {
...
   extraStyles: defaultCustomStyles,
-  tabIndex: "0",
+  tabIndex: 0,
```
Comment (YossiSaadi) [inline packages/core/src/components/Dropdown/Dropdown.tsx:696]: why this change?

File: packages/core/src/components/Dropdown/Dropdown.tsx [hunk 26]
```diff
@@ -397,13 +687,13 @@ Dropdown.defaultProps = {
   onInputChange: NOOP,
   searchable: true,
   options: [],
-  menuPlacement: Dropdown.menuPlacements.BOTTOM,
-  menuPosition: Dropdown.menuPositions.ABSOLUTE,
+  menuPlacement: DROPDOWN_MENU_PLACEMENT.BOTTOM,
+  menuPosition: DROPDOWN_MENU_POSITION.ABSOLUTE,
```
Comment (YossiSaadi) [inline packages/core/src/components/Dropdown/Dropdown.tsx:691]: Should be used as it was before, from the Dropdown's static props

File: packages/core/src/components/Dropdown/Dropdown.tsx [hunk 27]
```diff
@@ -397,13 +687,13 @@ Dropdown.defaultProps = {
   onInputChange: NOOP,
   searchable: true,
   options: [],
-  menuPlacement: Dropdown.menuPlacements.BOTTOM,
-  menuPosition: Dropdown.menuPositions.ABSOLUTE,
+  menuPlacement: DROPDOWN_MENU_PLACEMENT.BOTTOM,
```
Comment (YossiSaadi) [inline packages/core/src/components/Dropdown/Dropdown.tsx:690]: Should be used as it was before, from the Dropdown's static props

[the remark this answers is not in the document]
Comment (YossiSaadi) [reply packages/core/src/components/Dropdown/Dropdown.tsx:690]: everywhere where it has default of `undefined`, let's just not set default

File: packages/core/src/components/Dropdown/Dropdown.tsx [hunk 28]
```diff
@@ -20,14 +27,278 @@ import {
...
+   */
+  menuPosition?: DROPDOWN_MENU_POSITION;
+  /**
+   * If set to true, the dropdown will be in Right to Left mode
+   */
+  rtl?: boolean;
+  /**
+   * Set default selected value
+   */
+  defaultValue?: DropdownOption[];
```
Comment (YossiSaadi) [inline packages/core/src/components/Dropdown/Dropdown.tsx:172]:
```suggestion
  defaultValue?: DropdownOption[] | DropdownOption;
```

File: packages/core/src/components/Dropdown/Dropdown.tsx [hunk 29]
```diff
@@ -20,14 +27,278 @@ import {
...
+  /**
+   * custom menu render function
+   */
+  menuRenderer?: (...args: unknown[]) => unknown;
+  /**
+   * Default placement of the Dropdown menu in relation to its control. Use "auto" to flip the menu when there isn't enough space below the control.
+   */
+  menuPlacement?: DROPDOWN_MENU_PLACEMENT;
+
+  //  enum() PropTypes.oneOf(Object.values(DROPDOWN_MENU_PLACEMENT)),
```
Comment (YossiSaadi) [inline packages/core/src/components/Dropdown/Dropdown.tsx:159]: remove

File: packages/core/src/components/Dropdown/Dropdown.tsx [hunk 30]
```diff
@@ -20,14 +27,278 @@ import {
...
+  optionRenderer?: (option: DropdownOption) => JSX.Element;
+  /**
+   * custom value render function
+   */
+  valueRenderer?: (...args: unknown[]) => unknown;
+  ValueRenderer?: (...args: unknown[]) => unknown;
+  /**
+   * custom menu render function
+   */
+  menuRenderer?: (...args: unknown[]) => unknown;
```
Comment (YossiSaadi) [inline packages/core/src/components/Dropdown/Dropdown.tsx:153]: Sounds like all of those can be `ReactElement`, but let's leave if for now like this so we won't break existing behaviors

File: packages/core/src/components/Dropdown/Dropdown.tsx [hunk 31]
```diff
@@ -20,14 +27,278 @@ import {
...
+   */
+  searchable?: boolean;
+  /**
+   * The dropdown options
+   */
+  options?: DropdownOption[];
+  /**
+   * Text to display when there are no options
+   */
+  noOptionsMessage?: (() => string) | typeof NOOP;
```
Comment (YossiSaadi) [inline packages/core/src/components/Dropdown/Dropdown.tsx:128]: according to docs:
```
/* Text to display when there are no options */
  noOptionsMessage?: (obj: { inputValue: string }) => string | null;
```

File: packages/core/src/components/Dropdown/Dropdown.tsx [hunk 32]
```diff
@@ -20,14 +27,278 @@ import {
...
+   */
+  onInputChange?: (event: React.KeyboardEvent) => void;
+  /**
+   * If true, search in options will be enabled
+   */
+  searchable?: boolean;
+  /**
+   * The dropdown options
+   */
+  options?: DropdownOption[];
```
Comment (YossiSaadi) [inline packages/core/src/components/Dropdown/Dropdown.tsx:124]: @talkor let's monitor if it is breaking existing behaviors
if so:
> Everywhere where we use `DropdownOption`, we should be caution with it
We'll might need to replace to `OptionTypeBase` as a first step, and in the future make the Dropdown component accept a generic for `OptionType` that extends `OptionTypeBase`.

File: packages/core/src/components/Dropdown/Dropdown.tsx [hunk 33]
```diff
@@ -20,14 +27,278 @@ import {
...
+   * Called when selected value has changed
+   */
+  onChange?: (
+    option: DropdownOption | DropdownOption[],
+    event: SelectEvent | React.MouseEvent | React.KeyboardEvent
+  ) => void;
+  /**
+   * Called when the dropdown's input changes.
+   */
+  onInputChange?: (event: React.KeyboardEvent) => void;
```
Comment (YossiSaadi) [inline packages/core/src/components/Dropdown/Dropdown.tsx:116]: according to docs it does not get an event, but a string:
```
onInputChange?: (newValue: string, actionMeta: InputActionMeta) => void;
```

File: packages/core/src/components/Dropdown/Dropdown.tsx [hunk 34]
```diff
@@ -20,14 +27,278 @@ import {
...
+   */
+  onKeyDown?: (event: React.MouseEvent) => void;
+  /**
+   * Called when focused
+   */
+  onFocus?: (event: React.MouseEvent) => void;
+  /**
+   * Called when blurred
+   */
+  onBlur?: (event: React.MouseEvent) => void;
```
Comment (YossiSaadi) [inline packages/core/src/components/Dropdown/Dropdown.tsx:105]:
```suggestion
  onBlur?: FocusEventHandler
```

`FocusEventHandler` is imported from `react-select` alongside the other types

the `onBlur` is not a `MouseEvent`

File: packages/core/src/components/Dropdown/Dropdown.tsx [hunk 35]
```diff
@@ -20,14 +27,278 @@ import {
...
+   */
+  onMenuClose?: (event: React.MouseEvent) => void;
+  /**
+   * Called when key is pressed in the dropdown
+   */
+  onKeyDown?: (event: React.MouseEvent) => void;
+  /**
+   * Called when focused
+   */
+  onFocus?: (event: React.MouseEvent) => void;
```
Comment (YossiSaadi) [inline packages/core/src/components/Dropdown/Dropdown.tsx:101]:
```suggestion
  onFocus?: FocusEventHandler
```

`FocusEventHandler` is imported from `react-select` alongside the other types

the `onFocus` is not a `MouseEvent`

File: packages/core/src/components/Dropdown/Dropdown.tsx [hunk 36]
```diff
@@ -20,14 +27,278 @@ import {
...
+   */
+  onMenuOpen?: (event: React.MouseEvent) => void;
+  /**
+   * Called when menu is closed
+   */
+  onMenuClose?: (event: React.MouseEvent) => void;
+  /**
+   * Called when key is pressed in the dropdown
+   */
+  onKeyDown?: (event: React.MouseEvent) => void;
```
Comment (YossiSaadi) [inline packages/core/src/components/Dropdown/Dropdown.tsx:97]:
```suggestion
  onKeyDown?: KeyboardEventHandler
```

`KeyboardEventHandler` is imported from `react-select` alongside the other types

the `onKeyDown` is not a `MouseEvent`

**Review by YossiSaadi — COMMENTED** [review]
You did amazing @Mila2999 !
Thank you! 🙏🏼 

I went through mainly on the props, will go through the rest of stuff in another iteration

File: packages/core/types/packages.d.ts
Comment (YossiSaadi) [reply packages/core/types/packages.d.ts:1]: Ok I now see that in this version it doesn't include it, probably npm does not show the TS per version, but only for latest

File: packages/core/src/components/Dropdown/DropdownConstants.tsx (hunk 18 above)
[the remark this answers is not in the document]
Comment (YossiSaadi) [reply packages/core/src/components/Dropdown/DropdownConstants.tsx:32]: I rather having a `@ts-expect-error` or `@ts-ignore` on the place where `ThemeProvider` is used in the storybook rather than adding `Theme` here, I'm with @talkor on that, let's remove `Theme`

**Comment by YossiSaadi** [comment]
Hey @Mila2999 !
We would like to merge this PR as-is :)
we have fixes for mine and @talkor's comments in a different branch
you really did brilliant job!
would you mind fix the conflict with master so we can merge it?

**Review by talkor — APPROVED** [review]

**Comment by Mila2999** [comment]
Hi @YossiSaadi,

Thank you! I appreciate the feedback. I have rebased my code onto the master branch and resolved the conflicts in the latest commit. Sorry for any delays, as I'm currently on vacation, but I really would like to finalize this issue.

**Comment by talkor** [comment]
> Hi @YossiSaadi,
> 
> Thank you! I appreciate the feedback. I have rebased my code onto the master branch and resolved the conflicts in the latest commit. Sorry for any delays, as I'm currently on vacation, but I really would like to finalize this issue.

@Mila2999 thank you so much for the time you put into it! We really appreciate it! 🙏🏼
